### ISSUE TITLE
sourceRoot doesn't work for files in sub folders

### ISSUE THREAD

**Issue by SetTrend** [floridoo/gulp-sourcemaps#191]
I have stored my source files in a number of sub folders.

It looks like the `sourceRoot` parameter doesn't correctly handle those files. It doesn't consider the relative part from the sources' root to their actual location when creating the source map entry.

Please refer to this issue for reference:

https://github.com/Microsoft/vscode/issues/4705

**Comment by bryceg** [comment]
+1 I've been struggling with this also

**Comment by marcinant** [comment]
:+1:

**Comment by lexon0011** [comment]
Any solution??

**Comment by ckiely91** [comment]
This solution worked for me on Ubuntu but may not work on Windows, haven't tested that.

    .pipe(sourcemaps.write('.', {
            includeContent: false,
            sourceRoot: function(file) {
                return '../'.repeat(file.relative.split('/').length) + 'src';
            }
        }))

**Comment by lexon0011** [comment]
Thanks @ckiely91 - I only change `file.relative.split('/')` to `file.relative.split('\\')` so that it also works under windows.

**Comment by SetTrend** [comment]
Yes, but I prefer to keep in mind that this is only a workaround. No-one out there is aware of that quirk. It should get fixed soon.

**Comment by zhakhalov** [comment]
Hi
I've got same issue
Thanks @ckiely91. Your solution was very helpful. 
Here is my full task for "Node.js targeted" TS. Hope it would be useful...
```javascript
gulp.task('build', () => {
  const project = ts.createProject('tsconfig.json');
  return gulp.src('./app/**/*.ts')
    .pipe(sourcemaps.init())
    .pipe(ts(project))
    .js
    .pipe(sourcemaps.write('./', {
      includeContent: false,
      sourceRoot: file => `${'../'.repeat(file.relative.split(/^win/.test(process.platform) ? '\\' : '/').length)}app`
    }))
    .pipe(gulp.dest('./dist'));
});
```

I found following code in sources of plugin
```javascript
    if (typeof options.sourceRoot === 'function') {
      sourceMap.sourceRoot = options.sourceRoot(file);
    } else {
      sourceMap.sourceRoot = options.sourceRoot;
    }
```

Looks like sourceMap passed to output as is without, resolution of extra `../`

**Comment by ckiely91** [comment]
@zhakhalov, found that you can actually just use [path.sep](https://nodejs.org/api/path.html#path_path_sep) instead of `/^win/.test(process.platform) ? '\\' : '/'` to make it platform independent.

**Comment by terminalfrost** [comment]
I agree that the `sourceRoot` option does not work as expected when assigning a string value and source files are located in sub directories.

Here's a workaround using the `path` util:

```
const path = require('path');
// ...
sourceRoot: (file) => {
	return path.join(path.relative(path.join(outputDir, path.dirname(file.relative)), '.'), sourceDir);
}
```

**Comment by gms1** [comment]
workaround ( version 2.2.0 ) didn't work for me.
I guess
`var destSourceMapPath = path.join(file.cwd, options.destPath, mapFile);
`
should be changed to:
`
var destSourceMapPath = path.join(file.cwd, mapFile);
`
because mapFile already contains the destPath:
`
      var mapFile = path.join(destPath, file.relative) + '.map';
`

**Comment by jfstephe** [comment]
For whatever reason updating the sourceRoot as above didn't work for me so I updated the mapSources directly. Thanks to everyone for leading me to this!

mapSources: function(sourcePath, file) {
        let relativeLocation = path.join(path.relative(path.join(outputDir, path.dirname(file.relative)), '.'), 'src/');
        let relativeLocationToFile = path.join(relativeLocation, sourcePath);
        return relativeLocationToFile;
      }}))

**Comment by helloli** [comment]
The same problem when working with gulp-less. The less files in sub folders cannot pointing to less but css file.

**Comment by a8775** [comment]
+1

**Comment by doodlefolks** [comment]
+1 anyone working on this?

**Comment by demurgos** [comment]
Well, I had lots of trouble, trying to find the right combination to make it work. In the end I found a simple expression that worked for any combination of subdirectories, depth or relative location between the source and build directories:

```javascript
const writeSourceMapsOptions = {
  sourceRoot: (file) => {
    return path.posix.relative(path.posix.dirname(file.relative), "");
  },
};
```

and then just pass it to `.pipe(gulpSourceMaps.write(writeSourceMapsOptions))`.

This works both both Istambul reports and the Node debugger.

**Comment by nmccready** [comment]
@demurgos I wonder if you solution should become the default. Except, that it should work regardless of OS type. Feel free to add a PR with some specs.
